onnxruntime/core/providers/cpu/ml/svmclassifier.cc

```diff
@@ -125,120 +153,71 @@ Status SVMClassifier<T>::Compute(OpKernelContext* ctx) const {
-        float val = kernel_dot(x_data, current_weight_0, coefficients_, feature_count_ * j, feature_count_, get_kernel_type());
-        val += rho_[0];
-        scores.push_back(val);
-      }
     }
+
     if (proba_.size() > 0 && mode_ == SVM_TYPE::SVM_SVC) {
       //compute probabilities from the scores
-      std::vector<float> estimates;
-      std::vector<float> probsp2;
       int64_t num = class_count_ * class_count_;
-      for (int64_t m = 0; m < num; m++) {
-        probsp2.push_back(0.f);  //min prob
-      }
-      for (int64_t m = 0; m < class_count_; m++) {
-        estimates.push_back(0.f);  //min prob
-      }
+      std::vector<float> probsp2(num, 0.f);
+      std::vector<float> estimates(class_count_, 0.f);
       int64_t index = 0;
-      for (int64_t i = 0; i < class_count_; i++) {
-        for (int64_t j = i + 1; j < class_count_; j++) {
+      int64_t p1, p2;
+      for (int64_t i = 0; i < class_count_; ++i) {
+        p1 = i * class_count_ + i + 1;
+        p2 = (i + 1) * class_count_ + i;
+        for (int64_t j = i + 1; j < class_count_; ++j, ++index) {
           float val1 = sigmoid_probability(scores[index], proba_[index], probb_[index]);
           float val2 = std::max(val1, 1.0e-7f);
-          probsp2[i * class_count_ + j] = std::min(val2, 1 - 1.0e-7f);
-          probsp2[j * class_count_ + i] = 1 - probsp2[i * class_count_ + j];
-          index++;
+          val2 = std::min(val2, 1 - 1.0e-7f);
+          probsp2[p1] = val2;
+          probsp2[p2] = 1 - val2;
+          ++p1;
+          p2 += class_count_;
         }
       }
       multiclass_probability(class_count_, probsp2, estimates);
       //copy probabilities back into scores
       scores.resize(estimates.size());
-      for (int64_t k = 0; k < static_cast<int64_t>(estimates.size()); k++) {
-        scores[k] = estimates[k];
-      }
+      std::copy(estimates.begin(), estimates.end(), scores.begin());
+    // } else if (proba_.size() == 0) {
+    // Normalization will be changed in 0.21.
+    // See https://github.com/scikit-learn/scikit-learn/pull/10440
     }
-    int64_t maxvotes = 0;
+
+    double maxweight = 0;
     if (votes.size() > 0) {
-      for (int64_t k = 0; k < static_cast<int64_t>(votes.size()); k++) {
-        if (votes[k] > maxvotes) {
-          maxvotes = votes[k];
-          maxclass = k;
-        }
-      }
+      auto it_maxvotes = std::max_element(votes.begin(), votes.end());
+      maxclass = std::distance(votes.begin(), it_maxvotes);
     } else {
-      for (int64_t k = 0; k < static_cast<int64_t>(scores.size()); k++) {
-        if (scores[k] > maxweight) {
-          maxclass = k;
-          maxweight = scores[k];
-        }
-      }
+      auto it_maxweight = std::max_element(scores.begin(), scores.end());
+      maxclass = std::distance(scores.begin(), it_maxweight);
+      maxweight = *it_maxweight;
     }
-    //write top class
+
+    // write top class
+    // onnx specs expects one column per class.
```

Comment:
>specs [](start = 12, length = 5)

spec?

Reply:
Removed. The spec is ambiguous I think. One column per class if probabilities, it is more for raw scores.

Z : tensor(float)
Class scores (one per class per example), if prob_a and prob_b are provided they are probabilities for each class, otherwise they are raw scores.